src/detect.c

```diff
@@ -1256,13 +1267,17 @@ static DetectTransaction GetDetectTx(const uint8_t ipproto, const AppProto alpro
         detect_flags = 0;
     }
     if (detect_flags & APP_LAYER_TX_INSPECTED_FLAG) {
-        SCLogDebug("%"PRIu64" tx already fully inspected for %s. Flags %016"PRIx64,
-                tx_id, flow_flags & STREAM_TOSERVER ? "toserver" : "toclient",
-                detect_flags);
-        DetectTransaction no_tx = { NULL, 0, NULL, NULL, 0, 0, 0, 0, 0, };
+        SCLogDebug("%" PRIu64 " tx already fully inspected for %s. Flags %016" PRIx64, tx_id,
+                flow_flags & STREAM_TOSERVER ? "toserver" : "toclient", detect_flags);
         return no_tx;
     }
 
+    if (txd && txd->direction) {
+        if (txd->direction != (flow_flags & (STREAM_TOSERVER | STREAM_TOCLIENT))) {
+            return no_tx;
+        }
+    }
```

Comment:
Needs comment: Don't inspect unidirectional transactions in the reverse direction.